helpers/embeds.js

```diff
@@ -0,0 +1,65 @@
+import { html, nothing, render } from 'lit';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { map } from 'lit/directives/map.js';
+import { tryGetPluginByKey } from './plugins.js';
+
+const embedTypeAttributeName = 'data-d2l-embed-type';
+
+export async function createEmbedPlaceholder(embedType, props) {
+	const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', embedType);
+	if (!embedRendererPlugin) return;
+
+	const placeholderData = await embedRendererPlugin.getPlaceholder(props);
+	const contents = placeholderData.contents
+		? map(Object.entries(placeholderData.contents), ([id, content]) => html`<template data-d2l-embed-template-id=${id}>${content}</template>`)
+		: nothing;
+
+	const attributes = placeholderData.attributes ? JSON.stringify(placeholderData.attributes) : undefined;
+
+	return placeholderData.inline
+		? html`
+			<span data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(attributes)}">
+				${contents}
+			</span>
+		` : html`
+			<div data-d2l-embed-type="${embedType}" data-d2l-embed-props="${ifDefined(attributes)}">
+				${contents}
+			</div>
+		`;
+}
+
+export async function renderEmbeds(node, options) {
+	const placeholders = [...node.querySelectorAll(`div[${embedTypeAttributeName}], span[${embedTypeAttributeName}]`)];
+	if (placeholders.length === 0) return;
+
+	const processPlaceholder = async placeholder => {
+		const embedRendererPlugin = tryGetPluginByKey('d2l-html-embed-renderer', placeholder.dataset.d2lEmbedType);
+		if (!embedRendererPlugin) return;
+
+		const templates = [...placeholder.querySelectorAll('template[data-d2l-embed-template-id]')];
+		const processedTemplates = await Promise.all(templates.map(async template => {
+			const templateNode = template.content.cloneNode(true);
+			await renderEmbeds(templateNode, options);
+			return html`${templateNode}`;
+		}));
+
+		const processedTemplateContents = templates.reduce((acc, template, index) => {
+			acc[template.dataset.d2lEmbedTemplateId] = processedTemplates[index];
+			return acc;
+		}, {});
+
+		let props = placeholder.dataset.d2lEmbedProps;
+		try {
+			props = JSON.parse(placeholder.dataset.d2lEmbedProps);
+		} catch (e) { /* empty */ }
```

Comment:
This also feels like at the very least we'd want to `console.warn` it... otherwise the dev may be scratching their head for a while.

We could also `throw` in some of these cases that we're really not expecting, which would have the added bonus of being logged by the JS logging stuff (outside of content).

Reply:
Yeah, I wasn't 100% sure about this case in particular. It kind of depends on what we want to support, since the JSON parser is kind of picky. If we want to say "this must be JSON parse-able", then that seems fine to me, though it does force that on consumers (i.e. they couldn't set `data-d2l-embed-props="someStringProperty"`, for example).

Reply:
I'd be fine with a JSON-parse-able requirement. It kind of forces consumers to future-proof their data for a time when they might want more than 1 thing.

Reply:
Good point. In which case I think it's reasonable to just remove the try-catch and let the parser throw?